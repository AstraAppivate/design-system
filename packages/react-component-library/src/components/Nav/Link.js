import React from 'react'
import PropTypes from 'prop-types'

const Link = ({ active, className, href, label }) => (
  <a className={`${className}${active ? ' is-active' : ''}`} href={href}>
    {label}
  </a>
)

Link.propTypes = {
  active: PropTypes.bool,
  className: PropTypes.string,
  href: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
}

Link.defaultProps = {
  active: false,
  className: 'rn-nav__item',
}

export default Link
