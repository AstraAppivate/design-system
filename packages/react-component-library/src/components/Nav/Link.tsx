import React from 'react'

interface LinkProps {
  className?: string
  href: string
}

const Link: React.FC<LinkProps> = ({
  children,
  className = 'rn-nav__item',
  href,
}) => (
  <a className={className} href={href}>
    {children}
  </a>
)

Link.displayName = 'Link'

export default Link
