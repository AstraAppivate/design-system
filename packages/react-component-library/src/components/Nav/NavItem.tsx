/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState } from 'react'

interface NavItemProps {
  children?: any
  hasChildren: boolean
}

const NavItem: React.FC<NavItemProps> = ({ children, hasChildren = false }) => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <li
      className={`rn-nav__list-item ${hasChildren ? 'has-children' : ''} ${
        isOpen ? 'is-open' : ''
      }`}
      onClick={() => {
        setIsOpen(hasChildren && !isOpen)
      }}
    >
      {children}
    </li>
  )
}

NavItem.displayName = 'NavItem'

export default NavItem
