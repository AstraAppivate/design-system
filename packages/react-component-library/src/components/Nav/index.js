import React from 'react'
import PropTypes from 'prop-types'

const NavItem = navItem => (
  <a className="rn-nav__item" href={navItem.url}>
    {navItem.label}
  </a>
)

const Nav = ({ navItems, orientation }) => (
  <nav className={`rn-nav ${orientation}`}>{navItems.map(NavItem)}</nav>
)

Nav.propTypes = {
  /** The links to use within the nav, specified as an object with two items:
   * 'url': The link the nav item should go to,
   * 'label' : The text the nav item should display
   */
  navItems: PropTypes.arrayOf(
    PropTypes.shape({
      url: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  /** The type of navigation to use, can be one of 'horizontal' or 'vertical */
  orientation: PropTypes.string,
}

Nav.defaultProps = {
  orientation: 'vertical',
}

export default Nav
