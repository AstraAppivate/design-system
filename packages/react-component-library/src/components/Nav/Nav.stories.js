import React from 'react'

import { storiesOf } from '@storybook/react'
import { withKnobs } from '@storybook/addon-knobs'

import Nav from './index'

const stories = storiesOf('Nav', module)

stories.addDecorator(withKnobs)

const navItems = [
  {
    url: 'http://testurl.test',
    label: 'Home',
  },
  {
    url: 'http://testurl.test',
    label: 'Link 1',
  },
  {
    url: 'http://testurl.test',
    label: 'Link 2',
  },
  {
    url: 'http://testurl.test',
    label: 'Link 3',
  },
]

stories.add('Vertical', () => <Nav navItems={navItems} />)

stories.add('Horizontal', () => (
  <Nav navItems={navItems} orientation="horizontal" />
))
