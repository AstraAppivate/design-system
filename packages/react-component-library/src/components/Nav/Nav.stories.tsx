import React, { useState } from 'react'

import { storiesOf } from '@storybook/react'
import { withKnobs } from '@storybook/addon-knobs'

import { TriangleDown, TriangleUp } from '../../icons'
import CustomLink from '../CustomLink'
import Button from '../Button'
import Nav from './index'

const stories = storiesOf('Nav', module)

stories.addDecorator(withKnobs)

const navItems = [
  {
    href: 'http://testurl.test',
    label: 'Styles',
  },
  {
    href: 'http://testurl.test',
    label: 'Components',
  },
  {
    href: 'http://testurl.test',
    label: 'Patterns',
    active: true,
  },
  {
    href: 'http://testurl.test',
    label: 'Community',
  },
  {
    href: 'http://testurl.test',
    label: 'About',
  },
]

const navItemsWithChildren = [
  {
    href: 'http://testurl.test',
    label: 'Parent',
    children: [
      {
        href: 'http://testurl.test',
        label: 'Child 1',
      },
      {
        href: 'http://testurl.test',
        label: 'Child 2',
        children: [
          {
            href: 'http://testurl.test',
            label: 'Child 1',
          },
          {
            href: 'http://testurl.test',
            label: 'Child 2',
          },
        ],
      },
    ],
  },
  ...navItems,
]

stories.add('Vertical', () => <Nav navItems={navItemsWithChildren} />)

stories.add('Horizontal', () => (
  <Nav navItems={navItems} orientation="horizontal" />
))

stories.add('Sizes', () => (
  <div>
    <Nav navItems={navItems} orientation="horizontal" size="small" />
    <br />
    <Nav navItems={navItems} orientation="horizontal" />
    <br />
    <Nav navItems={navItems} orientation="horizontal" size="large" />
    <br />
    <Nav navItems={navItems} orientation="horizontal" size="xlarge" />

    <hr />

    <Nav navItems={navItems} orientation="vertical" size="small" />
    <br />
    <Nav navItems={navItems} orientation="vertical" />
    <br />
    <Nav navItems={navItems} orientation="vertical" size="large" />
    <br />
    <Nav navItems={navItems} orientation="vertical" size="xlarge" />
  </div>
))

const customNavItems = [
  {
    Component: CustomLink,
    to: 'http://testurl.test',
    label: 'Styles',
  },
  {
    Component: CustomLink,
    to: 'http://testurl.test',
    label: 'Components',
  },
  {
    Component: CustomLink,
    to: 'http://testurl.test',
    label: 'Patterns',
    active: true,
  },
  {
    Component: CustomLink,
    to: 'http://testurl.test',
    label: 'Community',
  },
  {
    Component: CustomLink,
    to: 'http://testurl.test',
    label: 'About',
  },
]

stories.add('Custom Item', () => <Nav navItems={customNavItems} />)

const PrimaryNav = () => {
  const [open, setOpen] = useState(true)

  const toggle = () => {
    setOpen(!open)
  }

  return (
    <div className="primary-nav">
      <Button
        className="primary-nav--button"
        onClick={toggle}
        icon={open ? <TriangleDown /> : <TriangleUp />}
      >
        Menu
      </Button>
      {open && <Nav navItems={navItems} orientation="horizontal" />}
    </div>
  )
}

stories.add('Primary Navigation', () => <PrimaryNav />)
