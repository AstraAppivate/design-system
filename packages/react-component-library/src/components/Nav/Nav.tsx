import React from 'react'
import { getKey } from '../../helpers'

import { Link } from '../Link'
import { NavItem } from './NavItem'
import logger from '../../utils/logger'
import { ComponentWithClass } from '../../common/ComponentWithClass'

interface NavProps extends ComponentWithClass {
  /**
   * Link component to use for each navigation item (custom implementation welcome).
   */
  LinkComponent?: any
  /**
   * Collection of navigation items to display in the list (JSON).
   */
  navItems: any[]
  /**
   * Orientation of the component.
   */
  orientation?: 'vertical' | 'horizontal'
  /**
   * Size of the component.
   */
  size?: 'small' | 'regular' | 'large' | 'xlarge'
}

function renderMenu(LinkComponent: any, navItems: any[]) {
  return (
    <ul className="rn-nav__list">
      {navItems.map((item) => {
        const { active, children, label, ...rest } = item

        const hasChildren: boolean = children && children.length > 0
        let subMenu: React.ReactNode | undefined

        if (hasChildren) {
          subMenu = renderMenu(LinkComponent, children)
        }

        return (
          <NavItem key={getKey('nav-item', label)} hasChildren={hasChildren}>
            <LinkComponent
              className={`rn-nav__item ${active ? 'is-active' : ''}`}
              {...rest}
            >
              {label}
            </LinkComponent>
            {subMenu}
          </NavItem>
        )
      })}
    </ul>
  )
}

export const Nav: React.FC<NavProps> = ({
  className = '',
  LinkComponent = Link,
  navItems,
  orientation = 'vertical',
  size = 'regular',
}) => {
  logger.warn(`Component \`Nav\` is deprecated`)

  return (
    <nav
      className={`rn-nav rn-nav--${orientation} rn-nav--${size} ${className}`}
      data-testid="nav"
    >
      {renderMenu(LinkComponent, navItems)}
    </nav>
  )
}

Nav.displayName = 'Nav'
