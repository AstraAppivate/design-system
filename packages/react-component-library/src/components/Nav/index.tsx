import React from 'react'
import uuid from 'uuid'

import NavItem from './NavItem'

// import '@royalnavy/css-framework/src/components/_nav.scss'

interface NavProps {
  className?: string
  navItems: any[]
  orientation?: 'vertical' | 'horizontal'
  size?: 'small' | 'regular' | 'large' | 'xlarge'
}

const Nav: React.FC<NavProps> = ({
  className = '',
  navItems,
  orientation = 'vertical',
  size = 'regular',
}) => (
  <nav className={`rn-nav rn-nav--${orientation} rn-nav--${size} ${className}`}>
    {navItems.map(item => (
      <NavItem key={uuid()} {...item} />
    ))}
  </nav>
)

export default Nav
