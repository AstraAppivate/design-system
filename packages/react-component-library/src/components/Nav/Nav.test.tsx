import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import toJson from 'enzyme-to-json'

import Nav from './index'

describe('Nav', () => {
  let component: ShallowWrapper
  let navItems: any[]
  let navItemsWithChildren: any[]

  describe('Given the nav is supplied navItems prop', () => {
    beforeEach(() => {
      navItems = [
        {
          href: 'http://testurl.test',
          label: 'Styles',
        },
        {
          href: 'http://testurl.test',
          label: 'Components',
        },
        {
          href: 'http://testurl.test',
          label: 'Patterns',
          active: true,
        },
        {
          href: 'http://testurl.test',
          label: 'Community',
        },
        {
          href: 'http://testurl.test',
          label: 'About',
        },
      ]

      navItemsWithChildren = [
        ...navItems,
        {
          href: 'http://testurl.test',
          label: 'Parent',
          children: [
            {
              href: 'http://testurl.test',
              label: 'Child 1',
            },
            {
              href: 'http://testurl.test',
              label: 'Child 2',
            },
          ],
        },
      ]
    })

    describe('and navItems is a flat collection of 5 items', () => {
      it('should render the correct number of list items', () => {
        component = shallow(<Nav navItems={navItems} />)
        expect(component.find('.rn-nav__list-item').length).toEqual(5)
      })

      it('snapshot: has same HTML structure', () => {
        const wrapper = shallow(
          <Nav navItems={navItems} orientation="vertical" />
        )
        expect(toJson(wrapper)).toMatchSnapshot()
      })
    })

    describe('and navItems is a nested collection of 5 items with children', () => {
      it('should render the correct number of list items', () => {
        component = shallow(<Nav navItems={navItemsWithChildren} />)
        expect(component.find('.rn-nav__list-item').length).toEqual(8)
      })
    })
  })

  describe('Given the nav is supplied orientation prop', () => {
    describe('and the value of the prop is vertical', () => {
      it('should apply the appropriate modifier class', () => {
        component = shallow(<Nav navItems={navItems} orientation="vertical" />)
        expect(component.find('.rn-nav').hasClass('rn-nav--vertical')).toBe(
          true
        )
      })
    })

    describe('and the value of the prop is horizontal', () => {
      it('should apply the appropriate modifier class', () => {
        component = shallow(
          <Nav navItems={navItems} orientation="horizontal" />
        )
        expect(component.find('.rn-nav').hasClass('rn-nav--horizontal')).toBe(
          true
        )
      })
    })

    describe('Given the nav is supplied size prop', () => {
      describe('and the value of the prop is small', () => {
        it('should apply the appropriate modifier class', () => {
          component = shallow(<Nav navItems={navItems} size="small" />)
          expect(component.find('.rn-nav').hasClass('rn-nav--small')).toBe(true)
        })
      })

      describe('and the value of the prop is regular', () => {
        it('should apply the appropriate modifier class', () => {
          component = shallow(<Nav navItems={navItems} size="regular" />)
          expect(component.find('.rn-nav').hasClass('rn-nav--regular')).toBe(
            true
          )
        })
      })

      describe('and the value of the prop is large', () => {
        it('should apply the appropriate modifier class', () => {
          component = shallow(<Nav navItems={navItems} size="large" />)
          expect(component.find('.rn-nav').hasClass('rn-nav--large')).toBe(true)
        })
      })

      describe('and the value of the prop is xlarge', () => {
        it('should apply the appropriate modifier class', () => {
          component = shallow(<Nav navItems={navItems} size="xlarge" />)
          expect(component.find('.rn-nav').hasClass('rn-nav--xlarge')).toBe(
            true
          )
        })
      })
    })
  })
})
