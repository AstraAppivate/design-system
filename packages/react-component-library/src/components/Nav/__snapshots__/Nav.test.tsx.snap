// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Given the nav is supplied navItems prop and navItems is a flat collection of 5 items snapshot: has same HTML structure 1`] = `
<Nav
  navItems={
    Array [
      Object {
        "href": "http://testurl.test",
        "label": "Styles",
      },
      Object {
        "href": "http://testurl.test",
        "label": "Components",
      },
      Object {
        "active": true,
        "href": "http://testurl.test",
        "label": "Patterns",
      },
      Object {
        "href": "http://testurl.test",
        "label": "Community",
      },
      Object {
        "href": "http://testurl.test",
        "label": "About",
      },
    ]
  }
  orientation="vertical"
>
  <nav
    className="rn-nav rn-nav--vertical rn-nav--regular "
  >
    <ul
      className="rn-nav__list"
    >
      <NavItem
        key="6"
      >
        <li
          className="rn-nav__list-item  "
          onClick={[Function]}
        >
          <Link
            className="rn-nav__item "
            href="http://testurl.test"
            label="Styles"
          >
            <a
              className="rn-nav__item "
              href="http://testurl.test"
            >
              Styles
            </a>
          </Link>
        </li>
      </NavItem>
      <NavItem
        key="7"
      >
        <li
          className="rn-nav__list-item  "
          onClick={[Function]}
        >
          <Link
            className="rn-nav__item "
            href="http://testurl.test"
            label="Components"
          >
            <a
              className="rn-nav__item "
              href="http://testurl.test"
            >
              Components
            </a>
          </Link>
        </li>
      </NavItem>
      <NavItem
        key="8"
      >
        <li
          className="rn-nav__list-item  "
          onClick={[Function]}
        >
          <Link
            active={true}
            className="rn-nav__item is-active"
            href="http://testurl.test"
            label="Patterns"
          >
            <a
              className="rn-nav__item is-active"
              href="http://testurl.test"
            >
              Patterns
            </a>
          </Link>
        </li>
      </NavItem>
      <NavItem
        key="9"
      >
        <li
          className="rn-nav__list-item  "
          onClick={[Function]}
        >
          <Link
            className="rn-nav__item "
            href="http://testurl.test"
            label="Community"
          >
            <a
              className="rn-nav__item "
              href="http://testurl.test"
            >
              Community
            </a>
          </Link>
        </li>
      </NavItem>
      <NavItem
        key="10"
      >
        <li
          className="rn-nav__list-item  "
          onClick={[Function]}
        >
          <Link
            className="rn-nav__item "
            href="http://testurl.test"
            label="About"
          >
            <a
              className="rn-nav__item "
              href="http://testurl.test"
            >
              About
            </a>
          </Link>
        </li>
      </NavItem>
    </ul>
  </nav>
</Nav>
`;
